metaflow/plugins/kubernetes/kubernetes_jobsets.py

```diff
@@ -0,0 +1,745 @@
+        if self._pod:
+            self._pod_name = self._pod["metadata"]["name"]
+            return self.id
+        return "jobset %s" % self._name
+
+    @property
+    def is_done(self):
+        def done():
+            return (
+                self._jobset_is_completed
+                or self._jobset_has_failed
+                or self._jobset_was_terminated
+            )
+
+        if not done():
+            # If not done, fetch newer status
+            self._jobset = self._fetch_jobset()
+            self._pod = self._fetch_pod()
+        return done()
+
+    @property
+    def status(self):
+        if self.is_done:
+            return "Jobset is done"
+
+        status = _construct_jobset_logical_status(self._jobset, control_pod=self._pod)
+        if status.status_unknown:
+            return "Jobset status is unknown"
+        if status.control_started:
+            if status.control_pod_status:
+                return f"Jobset is running: {status.control_pod_status}"
+            return "Jobset is running"
+        if status.all_jobs_are_suspended:
+            return "Jobset is waiting to be unsuspended"
+
+        return "Jobset waiting for jobs to start"
+
+    @property
+    def has_succeeded(self):
+        return self.is_done and self._jobset_is_completed
+
+    @property
+    def has_failed(self):
+        return self.is_done and self._jobset_has_failed
+
+    @property
+    def is_running(self):
+        if self.is_done:
+            return False
+        status = _construct_jobset_logical_status(self._jobset, control_pod=self._pod)
+        if status.control_started:
+            return True
+        return False
+
+    @property
+    def _jobset_was_terminated(self):
+        return _construct_jobset_logical_status(
+            self._jobset, control_pod=self._pod
+        ).jobset_was_terminated
+
+    @property
+    def is_waiting(self):
+        return not self.is_done and not self.is_running
+
+    @property
+    def reason(self):
+        # return exit code and reason
+        if self.is_done and not self.has_succeeded:
+            self._pod = self._fetch_pod()
+        elif self.has_succeeded:
+            return 0, None
+        status = _construct_jobset_logical_status(self._jobset, control_pod=self._pod)
+        if status.control_pod_failed:
+            return (
+                status.control_exit_code,
+                "control-pod failed [%s]" % status.control_pod_status,
+            )
+        elif status.worker_pods_failed:
+            return None, "Worker pods failed"
+        return None, None
+
+    @property
+    def _jobset_is_completed(self):
+        return _construct_jobset_logical_status(
+            self._jobset, control_pod=self._pod
+        ).jobset_finished
+
+    @property
+    def _jobset_has_failed(self):
+        return _construct_jobset_logical_status(
+            self._jobset, control_pod=self._pod
+        ).jobset_failed
+
+
+class TaskIdConstructor:
```

Comment:
This helps constructing all the task-ids from one place

Reply:
Excellent! Thanks for this implementation. Otherwise, it could get hard figuring out how/where the task ids get constructed.